build.zig

```diff
@@ -1,43 +1,121 @@
 // Zig version: 0.4.0
 
 const Builder = @import("std").build.Builder;
+const Step = @import("std").build.Step;
 const builtin = @import("builtin");
-const Array = @import("std").ArrayList;
+const std = @import("std");
+const ArrayList = std.ArrayList;
+const warn = std.debug.warn;
+const mem = std.mem;
+
+const src_files = [][]const u8{"kernel/kmain"};
+var debug = false;
+var arena: *std.mem.Allocator = undefined;
+
+fn concat(allocator: *std.mem.Allocator, str: []const u8, str2: []const u8) !std.Buffer {
+    var b = try std.Buffer.init(allocator, str);
+    try b.append(str2);
+    return b;
+}
 
 pub fn build(b: *Builder) void {
-    const kernel_out_dir = "bin/kernel";
-    const kernel_src = "src/kernel/";
+    var da = std.heap.DirectAllocator.init();
+    //defer da.deinit();
+    var aa = std.heap.ArenaAllocator.init(&da.allocator);
+    //defer aa.deinit();
+    arena = &aa.allocator;
 
-    var kernel = b.addExecutable("pluto.elf", kernel_src ++ "kmain.zig");
-    //kernel.addAssemblyFile(kernel_src ++ "start.s");
+    debug = b.option(bool, "debug", "build with debug symbols / make qemu wait for a debug connection") orelse false;
+    var build_path = b.option([]const u8, "build-path", "path to build to") orelse "bin";
+    var src_path = b.option([]const u8, "source-path", "path to source") orelse "src";
+    var target = b.option([]const u8, "target", "target to build/run for") orelse "x86";
+    const builtin_target = if (mem.eql(u8, target, "x86")) builtin.Arch.i386 else unreachable;
 
-    kernel.setOutputDir(kernel_out_dir);
-    kernel.setBuildMode(b.standardReleaseOptions());
-    kernel.setTarget(builtin.Arch.i386, builtin.Os.freestanding, builtin.Abi.gnu);
-    kernel.setLinkerScriptPath("link.ld");
+    const iso_path = concat(arena, build_path, "/pluto.iso") catch unreachable;
 
-    const run_objcopy = b.addSystemCommand([][]const u8 {
-        "objcopy", "-O", "binary", "-S", kernel.getOutputPath(), kernel_out_dir ++ "/pluto.bin",
-    });
-    run_objcopy.step.dependOn(&kernel.step);
+    var objects_steps = buildObjects(b, builtin_target, build_path, src_path);
+    var link_step = buildLink(b, builtin_target, build_path);
+    const iso_step = buildISO(b, build_path, iso_path.toSlice());
 
-    b.default_step.dependOn(&run_objcopy.step);
+    for (objects_steps.toSlice()) |step| b.default_step.dependOn(step);
+    b.default_step.dependOn(link_step);
+    for (iso_step.toSlice()) |step| b.default_step.dependOn(step);
 
-    const run_qemu = b.addSystemCommand([][]const u8 {
-        "qemu-system-i386",
-        "-display", "curses",
-        "-kernel", kernel.getOutputPath(),
-    });
+    buildRun(b, builtin_target, build_path, iso_path.toSlice());
+    buildDebug(b);
```

Comment:
Could add a buildTest that will rule tests, or have it as a option with the debug and release build that will run tests before running the OS